buf/registry/module/v1beta1/upload_service.proto

```diff
@@ -0,0 +1,93 @@
+// Copyright 2023 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+syntax = "proto3";
+
+package buf.registry.module.v1beta1;
+
+import "buf/registry/module/v1beta1/commit.proto";
+import "buf/registry/module/v1beta1/file.proto";
+import "buf/registry/module/v1beta1/label.proto";
+import "buf/registry/module/v1beta1/module.proto";
+import "buf/registry/module/v1beta1/resource.proto";
+import "buf/registry/module/v1beta1/vcs_commit.proto";
+import "buf/validate/validate.proto";
+
+option go_package = "buf.build/gen/go/bufbuild/registry/protocolbuffers/go/buf/registry/module/v1beta1";
+
+// Upload contents.
+service UploadService {
+  // Upload contents for given set of Modules.
+  //
+  // Content consists of the Files: .proto files, license files, and documentation files.
+  rpc Upload(UploadRequest) returns (UploadResponse) {}
+}
+
+message UploadRequest {
+  // A reference to upload content for.
+  //
+  // All references must have associated content.
+  // TODO: "Reference" isn't the best name.
+  message Reference {
+    // The Module of the reference.
+    ModuleRef module_ref = 1 [(buf.validate.field).required = true];
+    // The dependencies of the reference.
+    //
+    // This will include all transitive dependencies.
+    //
+    // Note that if a dependency should have Commits created for it, it should be specified
+    // as a Reference in the references list, and have associated Content. In this case, only
+    // the module component of the ResourceRef should be specified, and specifying any other
+    // part of the ResourceRef will result in an error.
+    //
+    // See the documentation on ResourceRef for resource resolution details.
+    //
+    // Resolution is as follows:
+    //   - If a Module is referenced, the Commit of the default Label is used.
+    //   - If a Label is referenced, the Commit of this Label is used.
+    //   - If a Commit is referenced, this Commit is used.
+    repeated ResourceRef dep_resource_refs = 2;
```

Comment:
Modules are referenced without commits when they are `References`:

```
In this case, only the module component of the ResourceRef should be specified.
```

This can either be a ModuleRef or a CommitRef, effectively. We figured we might as well allow LabelRefs for non-buf-CLI customers as well. The upload comment doesn't require resolution - it may have different callers than the buf CLI.